Max flow implementation with increased functionality and LCA improvements supporting weighted edges

cameroncuster opened the pull request:
#  New Maxflow Struct
The new maxflow implementation is described along with complexity in the source. The implementation supports increased functionality to the original Dinic implementation, but at the cost of slower run time.

## Utility
- find edges in the minimum cut
- flow of single edge
- send single flow across edge
- supports Dinic and Edmonds-Karp

## Results and Cross Comparisons
- the original maxflow struct was about 15x faster on this problem [Maximum Flow](https://open.kattis.com/problems/maxflow) where the only details modified where the struct used in the maxflow calculations
- Original imp (0.02s)
- New imp (0.3s)

![image](https://user-images.githubusercontent.com/73217097/127726604-f1750edf-1fbb-448f-b337-e6a79df63c4e.png)
Here are the cross comparisons of the flow implementations. Ignore the WA. The first one is original Dinic. The newest one is Edmonds-Karp with new struct, and the second to latest one is Dinic with the new struct.

## Testing
- The new struct is substantially slower, but has been tested on a variety of problems and proven it's utility
- Here are a few problems I have used for testing the new imp: [Distinct Routes](https://cses.fi/problemset/task/1711) - [sol](https://github.com/cameroncuster/CSESProblemSet/blob/main/graph_algorithms/distinct_routes.cpp), [Police Chase](https://cses.fi/problemset/task/1695) - [sol](https://github.com/cameroncuster/CSESProblemSet/blob/main/graph_algorithms/police_chase.cpp), [Download Speed](https://cses.fi/problemset/task/1694) - [sol](https://github.com/cameroncuster/CSESProblemSet/blob/main/graph_algorithms/download_speed.cpp)

# LCA Improvements and Utility
- supports weighted trees
- supports distance in edges and weight of path operations using LCA
- smaller include

## Testing
- Tested with this problem [The Shortest Statement](https://codeforces.com/contest/1051/problem/F) - [sol](https://codeforces.com/contest/1051/submission/124378359)

cameroncuster commented:
![image](https://user-images.githubusercontent.com/73217097/127726604-f1750edf-1fbb-448f-b337-e6a79df63c4e.png)
Here are the cross comparisons of the flow implementations. Ignore the WA. The first one is original Dinic. The newest one is Edmonds-Karp with new struct, and the second to latest one is Dinic with the new struct.

lrvideckis commented on graphs/lca.cpp:
```diff
@@ -1,53 +1,65 @@
+	lca(const vector<vector<pair<int, ll>>> &graph, int root) :
+		depth(graph.size()), dist(graph.size()) {//0 - based nodes
+			int n = graph.size();
+			Log = 1;
```
initializer list

cameroncuster replied:
For Log? - ammended - see submission below

lrvideckis commented on graphs/lca.cpp:
```diff
@@ -1,53 +1,65 @@
 		}
 	}
 
 	int kthPar(int node, int k) {
```
make const

cameroncuster replied:
done - submission below

lrvideckis replied:
I meant make the function const:
int kthPar(...) const {

cameroncuster replied:
okay sounds good

lrvideckis commented on graphs/lca.cpp:
```diff
@@ -1,53 +1,65 @@
 		return node;
 	}
 
 	int getLca(int x, int y) {
```
make const while were here

cameroncuster replied:
https://codeforces.com/contest/1051/submission/124398011

lrvideckis commented on graphs/lca.cpp:
```diff
@@ -1,53 +1,65 @@
-		memo.resize(n,vector<int>(Log));
-		dfs(root,root,graph);
-	}
+	vector<int> depth, dist;
```
dist should be a ll. Can you test with a problem where longs e required?

cameroncuster replied:
Yeah I see, I don't really know one, do you?

lrvideckis replied:
Hmm, couldn't find one. Can you test instead by writing naive, then running this against naive on random test data that requires longs?

lrvideckis replied:
And send me the files or results somehow

cameroncuster replied:
Yeah that'd work, might do this in the morning as I am in the midst of a random Web dev project

lrvideckis replied:
lol this problem should work https://codeforces.com/contest/1051/problem/F

lrvideckis replied:
(Don't forget to remove the #define int long long when testing

cameroncuster replied:
ah that makes sense. I was running with the macro so I guess that's why it worked.

cameroncuster replied:
https://codeforces.com/contest/1051/submission/124627472 here is the passing test. Can you make sure the code updated, I think it just pulls latest commit from my master until merged?

lrvideckis commented:
I’m against adding a slower max flow implementation. LMK if you want help speeding it up

lrvideckis reviewed (changes requested):
Please address the new comments

cameroncuster commented:
I do not know of a problem requiring longs that I can use to test it... otherwise all of your requests have been fulfilled for the LCA to my understanding.

lrvideckis commented on graphs/lca.cpp:
```diff
@@ -1,53 +1,64 @@
+	vector<int> depth, dist;
 
-	void dfs(int node, int par, const vector<vector<int>> &graph) {
+	// use weights of 1 for unweighted tree
```
Let's remove as we still have dist edges function

cameroncuster replied:
Remove the DFS?

lrvideckis replied:
remove the comment

lrvideckis replied:
actually nevermind, let's keep it

cameroncuster replied:
okay sounds good

lrvideckis reviewed (changes requested):
new comments

cameroncuster commented:
Do you want to just cherry-pick the LCA one, and then I can speed up the max flow at a later date?

lrvideckis commented:
> Do you want to just cherry-pick the LCA one, and then I can speed up the max flow at a later date?

sure, after comments are addressed

cameroncuster commented:
> > Do you want to just cherry-pick the LCA one, and then I can speed up the max flow at a later date?
> 
> sure, after comments are addressed

Have I not addressed all comments? Did the source not update on your end?

lrvideckis commented:
> > > Do you want to just cherry-pick the LCA one, and then I can speed up the max flow at a later date?
> > 
> > sure, after comments are addressed
> 
> Have I not addressed all comments? Did the source not update on your end?

- make functions const
- comment link of submission testing for ll’s

cameroncuster commented:
> > > > Do you want to just cherry-pick the LCA one, and then I can speed up the max flow at a later date?
> > > 
> > > 
> > > sure, after comments are addressed
> > 
> > 
> > Have I not addressed all comments? Did the source not update on your end?
> 
> * make functions const
> * comment link of submission testing for ll’s

I commented the link as a reply above ^,
and I made the functions const (really only two can be)

cameroncuster commented:
Here is the link again with the functions const https://codeforces.com/contest/1051/submission/124636795

cameroncuster commented:
> > > > Do you want to just cherry-pick the LCA one, and then I can speed up the max flow at a later date?
> > > 
> > > 
> > > sure, after comments are addressed
> > 
> > 
> > Have I not addressed all comments? Did the source not update on your end?
> 
> * make functions const
> * comment link of submission testing for ll’s

https://github.com/lrvideckis/Programming-Team-Code/pull/6#discussion_r680595236 here is a link to the comment containing the functions tested with longs

lrvideckis commented:
> Here is the link again with the functions const https://codeforces.com/contest/1051/submission/124636795

you have define int ll

lrvideckis commented:
but I saw the other submission, so we’re good

cameroncuster commented:
> > Here is the link again with the functions const https://codeforces.com/contest/1051/submission/124636795
> 
> you have define int ll

Yeah I know, I already did the other one, and this was easier.

cameroncuster commented:
> Here is the link again with the functions const https://codeforces.com/contest/1051/submission/124636795

Wait why did you 'dislike' this one? Because of the int to long long macro?

lrvideckis commented:
> > Here is the link again with the functions const https://codeforces.com/contest/1051/submission/124636795
> 
> Wait why did you 'dislike' this one? Because of the int to long long macro?

exactly - It doesn’t test for longs cuz of the macro

cameroncuster commented:
> > > Here is the link again with the functions const https://codeforces.com/contest/1051/submission/124636795
> > 
> > 
> > Wait why did you 'dislike' this one? Because of the int to long long macro?
> 
> exactly - It doesn’t test for longs cuz of the macro

Just to make sure we are keeping our testing standards the highest https://codeforces.com/contest/1051/submission/124637812

cameroncuster commented:
Thankfully I'm really like the LCA implementation we have now. I think it offers decent functionality for any problem I have seen (which are few)

lrvideckis commented on rangeDS/fenwickTree.cpp:
```diff
@@ -1,6 +1,16 @@
 struct fenwickTree {
 	vector<ll> bit;
 	fenwickTree(int n) : bit(n,0) {}
+	fenwickTree(const vector<int> & a) : bit(a.size()) {
```
Vector of ll

lrvideckis commented on rangeDS/fenwickTree.cpp:
```diff
@@ -1,6 +1,16 @@
+		bit[0] = a[0];
+		for (int i = 1; i < (int)a.size(); i++)
+			bit[i] = bit[i-1] + a[i];
+		for (int i = a.size() - 1; i > 0; i--) {
```
Cast a.size to an int, otherwise this will underflow in the case where a is empty

lrvideckis replied:
Then you'll get essentially a runtime error, it's a pretty nasty bug

lrvideckis commented on rangeDS/fenwickTree.cpp:
```diff
@@ -1,6 +1,16 @@
 	vector<ll> bit;
 	fenwickTree(int n) : bit(n,0) {}
+	fenwickTree(const vector<int> & a) : bit(a.size()) {
+		bit[0] = a[0];
```
This will RTE if a is empty
